Add support for nRF54L20 FLPR core

mstasiaknordic opened the pull request:
Building for `nrf54l20pdk/nrf54l20/cpuflpr` is now possible

mstasiaknordic committed:
dts: nordic: nrf54: add nRF54L20 FLPR core

Added support for nRF54L20 FLPR core in devicetree.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
soc: nordic: nrf54: add nRF54L20 FLPR core

Added support for nRF54L20 FLPR core in soc dir.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
modules: hal_nordic: nrfx: add nRF54L20 FLPR core

Added configuration files for nRF54L20 FLPR core.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
boards: nordic: add FLPR core to nRF54L20pdk

Added support for FLPR core in nRF54L20pdk board.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
snippets: add nRF54L20pdk FLPR core snippet

Add snippet to boot nRF54L20pdk FLPR from application core.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic commented on soc/nordic/nrf54l/Kconfig.defconfig.nrf54l20_enga_cpuflpr:
```diff
@@ -0,0 +1,11 @@
+if SOC_NRF54L20_ENGA_CPUFLPR
+
+config NUM_IRQS
+	default 303
```
271 interrupts + 32 VPR interrupts. Unless some of those are already included

mstasiaknordic commented:
@anangl Could you please take a look?

anangl commented on snippets/nordic-flpr/soc/nrf54l20_cpuapp.overlay:
```diff
@@ -0,0 +1,43 @@
+	status = "reserved";
+};
+
+&cpuapp_sram {
+	reg = <0x20000000 DT_SIZE_K(160)>;
+	ranges = <0x0 0x20000000 0x28000>;
+};
```
I don't quite get this. Why is this region shrunk if the released memory is not used in any way and `cpuflpr_sram_code_data` starts at 0x2006fc00, just as `cpuflpr_sram` does?

mstasiaknordic replied:
Probably a leftover from testing, removed it.

mstasiaknordic commented on boards/nordic/nrf54l20pdk/nrf54l20pdk_nrf54l20_cpuflpr.dts:
```diff
@@ -0,0 +1,70 @@
+&cpuflpr_sram {
+	status = "okay";
+	reg = <0x2006fc00 DT_SIZE_K(64)>;
+	ranges = <0x0 0x2006fc00 0x10000>;
```
Right, it was not needed. Removed it

mstasiaknordic commented on dts/common/nordic/nrf54l20.dtsi:
```diff
@@ -32,6 +36,16 @@
+			compatible = "nordic,vpr";
+			reg = <1>;
+			device_type = "cpu";
+			/* NRFX-6881: Restore to 128MHz when PDK becomes available. */
+			clock-frequency = <DT_FREQ_M(64)>;
```
Added in respect to https://github.com/zephyrproject-rtos/zephyr/pull/84599

mstasiaknordic commented on boards/nordic/nrf54l20pdk/nrf54l20pdk_nrf54l20-common.dtsi:
```diff
@@ -70,3 +70,15 @@
+	pinctrl-names = "default", "sleep";
+};
+
+&hfpll {
+	/* For now use 64 MHz clock for CPU and fast peripherals. */
+	clock-frequency = <DT_FREQ_M(64)>;
+};
```
Moved it from `nrf54l20_cpuapp_common.dtsi` so that both cores can use it

nordicjm commented on dts/common/nordic/nrf54l20.dtsi:
```diff
@@ -703,7 +741,13 @@
+				erase-block-size = <4096>;
+				write-block-size = <16>;
+			};
+			cpuflpr_rram: rram@1ed000 {
```
newline

nordicjm commented on boards/nordic/nrf54l20pdk/board.cmake:
```diff
@@ -1,7 +1,11 @@
 # SPDX-License-Identifier: Apache-2.0
 
-board_runner_args(jlink "--device=cortex-m33" "--speed=4000")
+if (CONFIG_BOARD_NRF54L20PDK_NRF54L20_CPUAPP)
```
no spaces between if and brackets in cmake

nordicjm commented on boards/nordic/nrf54l20pdk/doc/index.rst:
```diff
@@ -65,6 +65,13 @@ built, flashed, and debugged in the usual way. See
+to be built as multicore configuration with code snippet called ``vpr_launcher``
+for the application core.
+
+Enter the following command to compile ``hello_world`` for the FLPR core::
+ west build -p -b nrf54l20pdk/nrf54l20/cpuflpr --sysbuild -- -DSB_VPR_LAUNCHER=y
```
why not default this to on with a `Kconfig.sysbuild` file in the board folder?

mstasiaknordic replied:
I aligned it how it is defined for nRF54L15

nordicjm replied:
actually this isn't even right, the command there doesn't work and VPR_LAUNCHER is enabled automatically, so the whole thing can be removed. nrf54l15dk needs fixing too

mstasiaknordic replied:
Doesn't it work? In what way? Works fine on my side

nordicjm replied:
https://github.com/zephyrproject-rtos/zephyr/pull/86418

mstasiaknordic replied:
Okay, I agree with the correction

nordicjm commented on boards/nordic/nrf54l20pdk/nrf54l20pdk_nrf54l20-pinctrl.dtsi:
```diff
@@ -34,4 +34,26 @@
+			psels = <NRF_PSEL(UART_TX, 0, 0)>,
+				<NRF_PSEL(UART_RTS, 0, 2)>;
+		};
+		group2 {
```
newline

nordicjm commented on snippets/nordic-flpr/soc/nrf54l20_cpuapp.overlay:
```diff
@@ -0,0 +1,38 @@
+		reserved-memory {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			cpuflpr_code_partition: image@1ed000 {
```
newline

mstasiaknordic committed:
soc: nordic: enable west commands for nRF54L20 Enga cpuflpr

Enabled west flash and west erase for nRF54L20 PDK FLPR core.
Added missing reset qualifier for nRF54L20 and nRF54L09.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic commented:
@anangl Could you take a look?

anangl commented on soc/nordic/nrf54l/Kconfig.defconfig.nrf54l20_enga_cpuflpr:
```diff
@@ -0,0 +1,11 @@
+if SOC_NRF54L20_ENGA_CPUFLPR
+
+config NUM_IRQS
+	default 287
```
I think this should be 306 as VREGUSB IRQ is at index 289.

mstasiaknordic replied:
You are right, it got added in the meantime.

anangl commented on boards/nordic/nrf54l20pdk/board.yml:
```diff
@@ -4,3 +4,6 @@ board:
   vendor: nordic
   socs:
   - name: nrf54l20
+    variants:
+    - name: xip
+      cpucluster: cpuflpr
```
This is incorrect. There's no XIP variant defined.

mstasiaknordic replied:
I agree that XIP variant was not defined, but the naming would keep it consistent with other targets. I renamed it to `cpuflpr` for now

anangl replied:
I think I don't quite get what you mean by naming consistent with other targets any what's the point in that renaming you've done. Why not just remove the lines indicated above?

mstasiaknordic replied:
All flpr/ppr variants are named `xip` in `board.yml`. However, if getting rid of the cpuflpr variant description is legal, then I agree that it would be best

anangl replied:
> All flpr/ppr variants are named `xip` in `board.yml`.

Not quite. For other flpr and ppr targets we have two options: xip variant and non-xip. Here we have only the non-xip one.

mstasiaknordic replied:
Okay, I understood that variant generally refers to flpr/ppr target and thus has to be explicitly named in the `board.yml`, no matter whether it is xip or non-xip. Thanks for clarifying! In that case, I will remove it.

nordicjm commented on boards/nordic/nrf54l20pdk/nrf54l20pdk_nrf54l20_cpuflpr_defconfig:
```diff
@@ -0,0 +1,19 @@
+# Execute from SRAM
+CONFIG_XIP=n
+
+CONFIG_RISCV_ALWAYS_SWITCH_THROUGH_ECALL=y
```
why isn't this selected by Kconfig?

anangl commented on soc/nordic/soc.yml:
```diff
@@ -233,12 +235,17 @@ runners:
               - nrf54l05/cpuapp
               - nrf54l05/cpuflpr
+          - qualifiers:
+              - nrf54l09/cpuapp
```
Unintentional change, I guess.

mstasiaknordic committed:
bluetooth: controller: remove select of ENTROPY_NRF_CRACEN_CTR_DRBG

ENTROPY_NRF_CRACEN_CTR_DRBG symbol is based on devicetree
node with compatible nordic,nrf-cracen-ctrdrbg. It does not
have to be selected explicitly.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic commented:
Bluetooth controller Kconfig had to be modified because of `tests/bluetooth/classic/sdp_c`  failing on nRF54L20. It selected `ENTROPY_NRF_CRACEN_CTR_DRBG` symbol based on `SOC_COMPATIBLE_NRF54LX`, however, not all nRF54L targets have a `nordic,nrf-cracen-ctrdrbg` node defined.

cc @aescolar

aescolar commented:
Sidecomment:
> Bluetooth controller Kconfig had to be modified because of `tests/bluetooth/classic/sdp_c` failing on nRF54L20

Note this test is incorrectly being run in platforms which do not support it: 
https://discord.com/channels/720317445772017664/1014241011989487716/1351257906905088132

In any case, the removal of the select seems correct to me.
